### ISSUE TITLE
arm_post_processor_determine_calls() yields a false-positive result on call detection

### ISSUE THREAD

**Issue by danse-macabre** [angr/pyvex#162]
Consider the following disassembly excercept from [libnsl.so.1](https://github.com/angr/pyvex/files/2450124/libnsl.so.1.zip):

```
.text:0000B808                 STMFD   SP!, {R4,LR}
.text:0000B80C                 LDR     LR, =($_GLOBAL_OFFSET_TABLE_ - 0xB824)
.text:0000B810                 SUB     SP, SP, #0x28
.text:0000B814                 SUBS    R12, R1, #0
.text:0000B818                 STR     R0, [SP,#0x30+var_14]
.text:0000B81C                 ADD     LR, PC, LR      ; $_GLOBAL_OFFSET_TABLE_
.text:0000B820                 BEQ     loc_B874
```

and the corresponding IRSB (with opt_level set to 1):

```
IRSB {
   ...
   11 | ------ IMark(0xb80c, 4, 0) ------
   12 | t4 = LDle:I32(0x0000b8a8)
   13 | ------ IMark(0xb810, 4, 0) ------
   ...
   28 | ------ IMark(0xb81c, 4, 0) ------
   29 | t19 = Add32(0x0000b824,t4)
   30 | PUT(lr) = t19
   31 | PUT(pc) = 0x0000b820
   32 | ------ IMark(0xb820, 4, 0) ------
   33 | t42 = CmpEQ32(t8,0x00000000)
   34 | t41 = 1Uto32(t42)
   35 | t43 = 32to1(t41)
   36 | if (t43) { PUT(pc) = 0xb874; Ijk_Call }
   NEXT: PUT(pc) = 0x0000b824; Ijk_Boring
}
```

It seems that the post-processor wrongfully detects a call due to LR was assigned a value that somehow includes a PC.

Here is a script that reproduces the issue:
```python
import angr
p = angr.Project("libnsl.so.1")
b = p.factory.block(0x40b808)
assert all((j == 'Ijk_Boring' for j in b.vex.constant_jump_targets_and_jumpkinds.values()))
```

Here is the environment I'm working with:
```sh
angr environment report
=============================
Date: 2018-10-05 14:52:10.253721
Running in virtual environment at /home/user/.virtualenv/angr-dev
Platform: linux-x86_64
Python version: 3.5.3 (fdd60ed87e94, Apr 24 2018, 06:10:04)
[PyPy 6.0.0 with GCC 6.2.0 20160901]
######## angr #########
Python found it in /home/user/angr-dev/angr/angr
Pip version angr 8.18.10.1
Git info:
	Current commit 63b6b4eeb456c6c27fd96a839d5c601b8b99914b from branch master
	Checked out from remote origin: ssh://git@github.com/angr/angr
######## ailment #########
Python found it in /home/user/angr-dev/ailment/ailment
Pip version ailment 8.18.10.1
Git info:
	Current commit 966d5c57dd04f2e43f73b165ddca18166b249f85 from branch master
	Checked out from remote origin: ssh://git@github.com/angr/ailment
######## cle #########
Python found it in /home/user/angr-dev/cle/cle
Pip version cle 8.18.10.1
Git info:
	Current commit c56152f6b8e199bd897b80115e0a9cf43bb70f5d from branch master
	Checked out from remote origin: ssh://git@github.com/angr/cle
######## pyvex #########
Python found it in /home/user/angr-dev/pyvex/pyvex
Pip version pyvex 8.18.10.1
Git info:
	Current commit 14169efac878bad0b10d9e7333efbb49c45921d0 from branch master
	Checked out from remote origin: ssh://git@github.com/angr/pyvex
######## claripy #########
Python found it in /home/user/angr-dev/claripy/claripy
Pip version claripy 8.18.10.1
Git info:
	Current commit f4f2cbe47ddfe1c2cda9725846b3e9562c30ca92 from branch master
	Checked out from remote origin: ssh://git@github.com/angr/claripy
######## archinfo #########
Python found it in /home/user/angr-dev/archinfo/archinfo
Pip version archinfo 8.18.10.1
Git info:
	Current commit 0b8bf512c88ecfef6ac7bfe78fb9f263f88e54e5 from branch master
	Checked out from remote origin: ssh://git@github.com/angr/archinfo
######## ana #########
Python found it in /home/user/angr-dev/ana/ana
Pip version ana 0.5
Git info:
	Current commit 48657dc6500e33f5818b98003f1f179bfe03e1cc from branch master
	Checked out from remote origin: https://git:@github.com/zardus/ana
######## z3 #########
Python found it in /home/user/.virtualenv/angr-dev/site-packages/z3
Pip version not found!
Couldn't find git info
######## unicorn #########
Python found it in /home/user/.virtualenv/angr-dev/site-packages/unicorn
Pip version unicorn 1.0.1
Couldn't find git info

**Comment by subwire** [comment]
Yikes, isn't ARM awesome!? :/
We have a bunch of cases that look very similar that definitely _are_ calls that the compiler builds up by setting LR to a constant and then doing a jump.
Prior to angr 8, we had the setting of LR bounded by being the second-to-last instruction only, but I found some stuff compiled with armcc that violates this (ugh that compiler's output is gross!)

I think the trick will have to be the check as we have it + looking at where the jump actually goes.
Could you provide the rest of that function's asm?

**Comment by danse-macabre** [comment]
> I think the trick will have to be the check as we have it + looking at where the jump actually goes.

I think that it's not the case. The problem here is that the current post-processing code has failed to correctly trace the value of LR, which is definitely *not* pointing to the next IRSB address by the end of the current block. This is probably due to `Ist_Load` are not handled well inside the post-processing code.

I also would strongly suggest against "looking at where the jump actually goes," since this should never ever be done on behalf of the VEX lifter. I mean, the lifter is just a tool that lifts bytes, and it absolutely does _not_ need to know anything about functions or any related stuff. Anyway, you didn't want to fix the issue by looking whether the jump target is a function start, did you? :)

> Could you provide the rest of that function's asm?

Yep, here it is:

```
.text:0000B808 10 40 2D E9                 STMFD   SP!, {R4,LR}
.text:0000B80C 94 E0 9F E5                 LDR     LR, =($_GLOBAL_OFFSET_TABLE_ - 0xB824)
.text:0000B810 28 D0 4D E2                 SUB     SP, SP, #0x28
.text:0000B814 00 C0 51 E2                 SUBS    R12, R1, #0
.text:0000B818 1C 00 8D E5                 STR     R0, [SP,#0x30+var_14]
.text:0000B81C 0E E0 8F E0                 ADD     LR, PC, LR      ; $_GLOBAL_OFFSET_TABLE_
.text:0000B820 13 00 00 0A                 BEQ     loc_B874
.text:0000B824 80 10 9F E5                 LDR     R1, =(_xdr_nis_name - 0x1D000)
.text:0000B828 80 00 9F E5                 LDR     R0, =(_xdr_nis_error - 0x1D000)
.text:0000B82C 01 30 8E E0                 ADD     R3, LR, R1      ; _xdr_nis_name
.text:0000B830 00 E0 8E E0                 ADD     LR, LR, R0      ; _xdr_nis_error
.text:0000B834 1C 40 8D E2                 ADD     R4, SP, #0x30+var_14
.text:0000B838 0C 00 A0 E1                 MOV     R0, R12
.text:0000B83C 04 E0 8D E5                 STR     LR, [SP,#0x30+var_2C]
.text:0000B840 00 C0 A0 E3                 MOV     R12, #0
.text:0000B844 24 E0 8D E2                 ADD     LR, SP, #0x30+var_C
.text:0000B848 01 10 A0 E3                 MOV     R1, #1
.text:0000B84C 16 20 A0 E3                 MOV     R2, #0x16
.text:0000B850 00 40 8D E5                 STR     R4, [SP,#0x30+var_30]
.text:0000B854 08 E0 8D E5                 STR     LR, [SP,#0x30+var_28]
.text:0000B858 10 C0 8D E5                 STR     R12, [SP,#0x30+var_20]
.text:0000B85C 0C C0 8D E5                 STR     R12, [SP,#0x30+var_24]
.text:0000B860 8F F6 FF EB                 BL      __do_niscall2
.text:0000B864
.text:0000B864             loc_B864                                ; CODE XREF: nis_mkdir+9C↓j
.text:0000B864 00 00 50 E3                 CMP     R0, #0
.text:0000B868 24 00 9D 05                 LDREQ   R0, [SP,#0x30+var_C]
.text:0000B86C 28 D0 8D E2                 ADD     SP, SP, #0x28
.text:0000B870 10 80 BD E8                 LDMFD   SP!, {R4,PC}
.text:0000B874             ; ---------------------------------------------------------------------------
.text:0000B874
.text:0000B874             loc_B874                                ; CODE XREF: nis_mkdir+18↑j
.text:0000B874 34 30 9F E5                 LDR     R3, =(_xdr_nis_error - 0x1D000)
.text:0000B878 2C 20 9F E5                 LDR     R2, =(_xdr_nis_name - 0x1D000)
.text:0000B87C 16 10 A0 E3                 MOV     R1, #0x16
.text:0000B880 02 20 8E E0                 ADD     R2, LR, R2      ; _xdr_nis_name
.text:0000B884 03 E0 8E E0                 ADD     LR, LR, R3      ; _xdr_nis_error
.text:0000B888 00 E0 8D E5                 STR     LR, [SP,#0x30+var_30]
.text:0000B88C 1C 30 8D E2                 ADD     R3, SP, #0x30+var_14
.text:0000B890 24 E0 8D E2                 ADD     LR, SP, #0x30+var_C
.text:0000B894 04 E0 8D E5                 STR     LR, [SP,#0x30+var_2C]
.text:0000B898 0C C0 8D E5                 STR     R12, [SP,#0x30+var_24]
.text:0000B89C 08 C0 8D E5                 STR     R12, [SP,#0x30+var_28]
.text:0000B8A0 56 F6 FF EB                 BL      __do_niscall
.text:0000B8A4 EE FF FF EA                 B       loc_B864
.text:0000B8A4             ; End of function nis_mkdir
.text:0000B8A4
.text:0000B8A4             ; ---------------------------------------------------------------------------
.text:0000B8A8 DC 17 01 00 off_B8A8        DCD $_GLOBAL_OFFSET_TABLE_ - 0xB824
.text:0000B8A8                                                     ; DATA XREF: nis_mkdir+4↑r
.text:0000B8AC 50 DC FE FF off_B8AC        DCD _xdr_nis_name - 0x1D000
.text:0000B8AC                                                     ; DATA XREF: nis_mkdir+1C↑r
.text:0000B8AC                                                     ; nis_mkdir+70↑r
.text:0000B8B0 58 DB FE FF off_B8B0        DCD _xdr_nis_error - 0x1D000
.text:0000B8B0                                                     ; DATA XREF: nis_mkdir+20↑r
.text:0000B8B0                                                     ; nis_mkdir:loc_B874↑r
```

**Comment by subwire** [comment]
This will (eventually) be fixed in #166
